rbac/default-role-manager/role_manager.go

```diff
@@ -288,6 +321,12 @@ func (roles *Roles) createRole(name string, matchingFunc MatchingFunc) *Role {
 				role1, _ := roles.LoadOrStore(key.(string), newRole(key.(string)))
 				role.(*Role).addRole(role1.(*Role))
 			}
+			// name serve as pattern
+			if matchingFunc(key.(string), name) && name != key.(string) {
+				role1, _ := roles.LoadOrStore(key.(string), newRole(key.(string)))
+				role1.(*Role).addRole(role.(*Role))
+			}
+
 			return true
 		})
 	}
```

Comment:
line 317 - 332 should be deleted, because the links will be created in `AddLink`.

I already tried to delete those lines and as a result a test case fails.
Output:
```
--- FAIL: TestImplicitRoleAPI (0.00s)
    rbac_api_test.go:280: Permissions for  alice :  [[alice data1 read]]
    rbac_api_test.go:281: Permissions for  bob :  [[bob data2 write]]
    rbac_api_test.go:283: Implicit roles for  alice :  [admin data1_admin data2_admin]
    rbac_api_test.go:284: Implicit roles for  bob :  []
    rbac_api_test.go:291: Implicit roles for  cathy :  [/book/1/2/3/4/5 pen_admin]
    rbac_api_test.go:291: Implicit roles for  cathy :  [/book/1/2/3/4/5 pen_admin] , supposed to be  [/book/1/2/3/4/5 pen_admin /book/* book_group]
    rbac_api_test.go:292: Roles for  cathy :  [/book/1/2/3/4/5 pen_admin]
FAIL
```
corresponding policy:
```
p, alice, /pen/1, GET
p, alice, /pen2/1, GET
p, book_admin, book_group, GET
p, pen_admin, pen_group, GET

g, alice, book_admin
g, bob, pen_admin

g, cathy, /book/1/2/3/4/5
g, cathy, pen_admin

g2, /book/*, book_group

g2, /book/:id, book_group
g2, /pen/:id, pen_group

g2, /book2/{id}, book_group
g2, /pen2/{id}, pen_group
```
In my opinion the test case is wrong. 
I would expect `cathy` to have only the roles `/book/1/2/3/4/5` and `pen_admin`, because the policy specifies separated `g` functions.

The question is why did this test case not fail in the first place.

Reply:
In `GetImplicitRolesForUser`, it will traverse all `rbac` system, like `g`, `g2`. It will add the roles from the last `GetRoles()` to the `q` array, and the roles from the `q` array will be passed to the GetRoles as arguments. You could see `rbac_api.go line 171-188`

Reply:
You are right. Maybe we should create a new issue to fix `GetImplicitRolesForUser`.

But with mentioned lines in place it is still possible to produce unwanted links.
```go
_ = rm.Clear()

_ = rm.AddLink("u1", "g\\d+")
_, _ = rm.GetRoles("g1")
_, _ = rm.GetRoles("g2")

testRole(t, rm, "u1", "g1", true)
testRole(t, rm, "u1", "g1", true)
testRole(t, rm, "g1", "g2", false)
testRole(t, rm, "g2", "g1", false)
```
output:
```
role_manager_test.go:304: u1, g1: true
role_manager_test.go:305: u1, g2: true
role_manager_test.go:306: g1, g2: true
role_manager_test.go:306: g1 < g2: true, supposed to be false
role_manager_test.go:307: g2, g1: true
role_manager_test.go:307: g2 < g1: true, supposed to be false
```

Reply:
Yeah, maybe we should delete those lines in `role_manager.go`, and create a new issue to discuss these test case